[DataLoader] Close open in DataPipe streams on best effort basis

VitalyFedyunin opened the pull request:
Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78952

Differential Revision: [D37489935](https://our.internmc.facebook.com/intern/diff/D37489935)

VitalyFedyunin committed:
Debug pipes overflow

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Debug pipes overflow"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Debug pipes overflow"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

NivekT reviewed (commented):
What happens if some does `fork` then `filter` (or `zip`)? Would this close the stream held by the other ChildDataPipe?

```python
stream_dp = ...
cdp1, cdp2 = stream_dp.fork(2)
cdp2 = cdp2.filter(...)
list(cdp1.readlines())
```
I think we have to increment the counter when `fork` is used, and only close a stream when there is no other reference to it.

VitalyFedyunin commented:
> What happens if some does `fork` then `filter` (or `zip`)? Would this close the stream held by the other ChildDataPipe?
> 
> ```python
> stream_dp = ...
> cdp1, cdp2 = stream_dp.fork(2)
> cdp2 = cdp2.filter(...)
> list(cdp1.readlines())
> ```
> 
> I think we have to increment the counter when `fork` is used, and only close a stream when there is no other reference to it.

Ideally, we should `warn` and RuntimeError if stream is forked. IMO there is no good way to allow reading abstract buffer twice.

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -510,9 +511,18 @@ def __init__(self, *datapipes: IterDataPipe):
 
     def __iter__(self) -> Iterator[Tuple[T_co]]:
-        for data in zip(*self.datapipes):
+        iterators = [iter(datapipe) for datapipe in self.datapipes]
+        for data in zip(*iterators):
             yield data
 
+        unused = []
+        for iterator in iterators:
+            unused += list(iterator)
+
+        # TODO(VitalyFedyunin): This should be Exception or warning when torchdata.debug is enabled
+        for item in unused:
+            StreamWrapper.close_streams(item)
```
I am not sure about the behavior of generator functions. When an not fully consumed iterator is deconstructed, I think the iterator would be destroyed. Technically, these `StreamWrapper` in `unused` variable should not even be created.
Do we need to handle this case?

NivekT replied:
I think there are two options here:
1. Do `try...finally` within `__iter__`, where the `finally` clause does the clean up.
  - This should work regardless how/why the deconstruction of the iterator happens. 
I just tried this with `IterableWrapper`.
```python
    def __iter__(self):
        try:
            source_data = self.iterable
            if self.deepcopy:
                try:
                    source_data = copy.deepcopy(self.iterable)
                # For the case that data cannot be deep-copied,
                # all in-place operations will affect iterable variable.
                # When this DataPipe is iterated second time, it will
                # yield modified items.
                except TypeError:
                    warnings.warn(
                        "The input iterable can not be deepcopied, "
                        "please be aware of in-place modification would affect source data."
                    )
            for data in source_data:
                yield data
        finally:
            print("FINALLLY")
```
```python
dp = IterableWrapper(range(10))
it = iter(dp)
print(next(it))
print("About to create a new iterator")
it = iter(dp)
print("Created a new iterator")
```
```
0
About to create a new iterator
FINALLLY
Created a new iterator
```
2. We can put the clean up logic with `hook_iterator` or `reset` method of `IterDataPipe` whenever that is triggered.

VitalyFedyunin replied:
I like the idea of `finally:` it will produce readable code and can be placed as a development pattern

VitalyFedyunin replied:
Done

ejguan commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -157,13 +157,59 @@ class StreamWrapper:
     the wrapped file handler is closed when it's out of scope.
     '''
-    def __init__(self, file_obj):
+    session_streams: Dict[Any, int] = {}
```
We can use a `set` in this case.
Do you think we might also register a function to `atexit` to make sure all `session_streams` are closed?

VitalyFedyunin replied:
Oddly it doesn't allow to put self into Set. Switching back to Dict.

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37489935).

NivekT commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -216,13 +218,59 @@ class StreamWrapper:
+        if self.child_counter == 0:
+            self.close()
+        self.close_on_last_child = True
+
```
Do we need to change the implementation of `__del__` here to remove `self` from `StreamWrapper.session_streams`?

ejguan replied:
I agree on this comment

NivekT reviewed (commented):
1. I think we need some tests to reason through the common and complicated cases.
2. Question: when a buffer is cleared (say in `Shuffler`), do we need to attempt to call `close_streams`?

ejguan commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -216,13 +218,59 @@ class StreamWrapper:
+            if isinstance(v, dict):
+                for kk, vv in v.items():
+                    cls.close_streams(vv, depth=depth + 1)
+            elif isinstance(v, list) or isinstance(v, tuple):
```
nit: `isinstance(v, (list, tuple))`

ejguan commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -216,13 +218,59 @@ class StreamWrapper:
+            self.parent_stream.child_counter -= 1
+            if not self.parent_stream.child_counter and self.parent_stream.close_on_last_child:
+                self.parent_stream.close()
+        self.file_obj.close(*args, **kwargs)
```
```suggestion
        try:
            self.file_obj.close(*args, **kwargs)
        except AttributeError:
            pass
```

VitalyFedyunin commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -216,13 +218,59 @@ class StreamWrapper:
+                self.parent_stream.close()
+        self.file_obj.close(*args, **kwargs)
+
+    def autoclose(self):
+        '''
+        Marks Steam to close automatically as soon as all child streams are closed.
+        '''
+        if self.child_counter == 0:
+            self.close()
+        self.close_on_last_child = True
```
It will close stream if there is no active children or mark it to be closed automatically if there is at least one.

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures


Differential Revision: [D37489935](https://our.internmc.facebook.com/intern/diff/D37489935)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Close open in DataPipe streams on best effort basis"

Adding ability to:
- Track open StreamWrappers with `StreamWrapper.session_streams`
- Automatically close parent StreamWrapper (ex. torchdata tar is the parent and extracted file streams are children)
- Close streams in discarded by filtering structures


Differential Revision: [D37489935](https://our.internmc.facebook.com/intern/diff/D37489935)

[ghstack-poisoned]

VitalyFedyunin commented:
@pytorchbot merge -g
